exp/addons/controllers/clusterresourceset_controller.go

```diff
@@ -0,0 +1,357 @@
+		return ctrl.Result{}, err
+	}
+
+	for _, cluster := range clusters {
+		if err := r.ApplyClusterResourceSet(ctx, cluster, clusterResourceSet); err != nil {
+			// NOTE: The reason of just logging the error if applying resources are failed is to avoid retries in case resources are missing.
+			logger.Error(err, "Failed applying resources to cluster", "Cluster", cluster.Name)
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// getClustersByClusterResourceSetSelector fetches Clusters matched by the ClusterResourceSet's label selector that are in the same namespace as the ClusterResourceSet object.
+func (r *ClusterResourceSetReconciler) getClustersByClusterResourceSetSelector(ctx context.Context, clusterResourceSet *addonsv1.ClusterResourceSet) ([]*clusterv1.Cluster, error) {
+	clusterList := &clusterv1.ClusterList{}
+	if err := r.Client.List(ctx, clusterList, client.InNamespace(clusterResourceSet.Namespace), client.MatchingLabels(clusterResourceSet.Spec.ClusterSelector.MatchLabels)); err != nil {
+		return nil, errors.Wrap(err, "failed to list clusters")
+	}
+
+	clusters := []*clusterv1.Cluster{}
+	for i := range clusterList.Items {
+		m := &clusterList.Items[i]
+		if m.DeletionTimestamp.IsZero() {
+			clusters = append(clusters, m)
+		}
+	}
+	return clusters, nil
+}
+
+func (r *ClusterResourceSetReconciler) ApplyClusterResourceSet(ctx context.Context, cluster *clusterv1.Cluster, clusterResourceSet *addonsv1.ClusterResourceSet) error {
+	logger := r.Log.WithValues("ClusterResourceSet", clusterResourceSet.Name, "namespace", clusterResourceSet.Namespace)
+
+	logger.Info("Applying ClusterResourceSet to cluster", "Cluster", cluster.Name)
+
+	err := r.applyClusterResourceSet(ctx, cluster, clusterResourceSet)
+	if err != nil {
+		logger.Error(err, "Failed to apply resources")
+		return err
+	}
+	conditions.MarkTrue(clusterResourceSet, addonsv1.ResourcesAppliedCondition)
+
+	return nil
+}
+
+// applyClusterResourceSet applies resources in a ClusterResourceSet to a Cluster. Once applied, a record will be added to the
+// cluster's ClusterResourceSetBinding.
+// In ApplyOnce strategy, resources are applied only once to a particular cluster. ClusterResourceSetBinding is used to check if a resource is applied before.
+// It applies resources best effort and continue on scenarios like: unsupported resource types, failure during creation, missing resources.
+// TODO: If a resource already exists in the cluster but not applied by ClusterResourceSet, the resource will be updated ?
+func (r *ClusterResourceSetReconciler) applyClusterResourceSet(ctx context.Context, cluster *clusterv1.Cluster, clusterResourceSet *addonsv1.ClusterResourceSet) error {
+	logger := r.Log.WithValues("ClusterResourceSet", clusterResourceSet.Name, "namespace", clusterResourceSet.Namespace)
+
+	c, err := remote.NewClusterClient(context.Background(), r.Client, client.ObjectKey{Namespace: cluster.Namespace, Name: cluster.Name}, r.scheme)
```

Comment:
I think we want to pass in a ClusterCacheTracker when creating the ClusterResourceSetReconciler, and use the tracker's GetClient() method (this keeps us from creating clients over and over). cc @vincepri to confirm.